### ISSUE TITLE
APM Error: discarding invalid livedata message

### ISSUE THREAD

**Issue by rj-david** [meteor/meteor#11513]
- The version of Meteor showing the problem.
Meteor 2.3
- The last version of Meteor where the problem did _not_ occur, if applicable.
Meteor 2.2
- The operating system you're running Meteor on.
Ubuntu

We started receiving this error in APM since we upgraded to Meteor 2.3: `discarding invalid livedata message`

**Comment by StorytellerCZ** [comment]
Could be potentially related to handshake between client and server which was adjusted going to 2.3.

**Comment by brianlukoff** [comment]
We are seeing the same thing after upgrading to 2.3.

**Comment by rj-david** [comment]
Here is the change that added the debug message. It's possible that the cause to be happening even before 2.3
https://github.com/meteor/meteor/commit/d53ab4cd177413055402f7cf3db9936c209d935f

**Comment by StorytellerCZ** [comment]
We could potentially not report the discarded message, but I don't think that is a good approach. I will update the affected code to use the logging tool so that the debug message won't appear in production.

**Comment by filipenevola** [comment]
Hi @rj-david do you see an object after `discarding invalid livedata message`?

As you can see on my change we shouldn't receive a message object without a msg field on it so that is why we have this log. 

Maybe it has another fields inside this object and it could give us some hint where it is coming from.

And I did this change as it was breaking a test after some removal of deprecated code.

**Comment by filipenevola** [comment]
> We could potentially not report the discarded message, but I don't think that is a good approach. I will update the affected code to use the logging tool so that the debug message won't appear in production.

Changing the way we log messages is not going to solve this issue. We need to understand which message is being discarded.

**Comment by rj-david** [comment]
> Hi @rj-david do you see an object after `discarding invalid livedata message`?

Monti APM only showed `[object Object]` unfortunately.

This happened only 4 times during the past week.

**Comment by tmihelcich** [comment]
I just started receiving this message during development on a project that is using Meteor 1.4.3.2. Here is what I see:

`discarding invalid livedata message { testMessageOnConnect: true }`

It happens just after I save changes to file and the app reloads.

**Comment by PolarisWT** [comment]
I can confirm same message using Meteor 2.2 on local environment.   Local server running for past day or so without any updates in between.  Just started showing this log on client updates

```console
=> Client modified -- refreshingdiscarding invalid livedata message { testMessageOnConnect: true }
=> Client modified -- refreshing (x2)discarding invalid livedata message { testMessageOnConnect: true }
=> Client modified -- refreshing (x3)discarding invalid livedata message { testMessageOnConnect: true }
...
```

**Comment by dpatte** [comment]
i see this on my local environment in 1.8.1 - exact same message as PolarisWT. Started a few days ago.

**Comment by kakadais** [comment]
I've seen this too at METEOR@1.10.2 on local & server env.
I don't think I've seen this before on the same project.

**Comment by rastko909** [comment]
Running on 1.22.0

>  Client modified -- refreshingdiscarding invalid livedata message { testMessageOnConnect: true }

**Comment by AstonBraham** [comment]
Same on my local environment. Meteor 2.2.1

**Comment by filipenevola** [comment]
Hi, this message should be sent only in test mode, it was created just to keep an old test that was checking if Meteor was connected still running.

**Comment by tonymckendry** [comment]
So are we just going to ignore this new change? To an earlier poster's point, I am seeing the message every time I make a change to a file in the project while the server is running. This should be removed or made configurable.

**Comment by filipenevola** [comment]
I can say for sure that you can ignore this message, it's not causing any problem.

That was the reason I asked the object after the log, so I was sure what was the root cause.

We are going to fix it soon 😉

**Comment by tomtom87** [comment]
I didn't update and it's just suddenly started today, compiling on Mac - nothing was changed with Node version or Meteor it just appeared when I built today

**Comment by filipenevola** [comment]
I believe the root cause of this message being displayed is that the `ddp-client` is not up-to-date.

Could you check your versions?

I don't see this message in the server console, as expected, with these versions:
- `ddp-client@2.5.0`
- `ddp-server@2.4.0` 

But I do see with these versions:
- `ddp-client@2.3.3`
- `ddp-server@2.3.2`

Please update your packages and if the message persist let me know and we can re-open the issue.

> Again, this issue is not going to cause any problem, the way to ignore this message in the client was modified so if you have a mismatch between the versions of `ddp-client` and `ddp-server` the message is going to be displayed but this is not going to cause any problems at all.

**Comment by kakadais** [comment]
ddp-client@2.3.3 in my case.
It sounds figured the reason out but still, I don't get why the message appeared all of a sudden without any update.

**Comment by tomtom87** [comment]
Yeah me too, seems really weird that can appear without update. 

Any chance that this is some form of tracking being added that was bugged? I didn’t realize meteor can push out updates without the developer requesting them. Very interesting thanks for the feedback. 

>

**Comment by kakadais** [comment]
That's my point.

I think one of the best points of Meteor is 'Manage & Keep the project independently'.
It still works great than the other platforms,
but this one is critical of the dependencies things.

**Comment by StorytellerCZ** [comment]
@kakadais from the OP it looks like there has been an update to Meteor 2.3 which removed some deprecated code. From the looks of it I failed to account for an impact of one of the removal would have, which was this message. It was fixed in the later version and then I forget that I pushed out a patch with it as it was just a one liner that came in together with something else. I have then moved onto other things and completely forgot about this. My apologies for the confusion.

@tomtom87 No this was not tracking or anything like it. Feel free to inspect the code and the changes. This message was always there, but the handling in hiding it was in a deprecated part of the code. The reason why you saw that message was the opposite of what you think. You did not request an update so you did not get a fix where the message was ignored as was the previous behavior.

**Comment by bhushitjoshipura** [comment]
I am using Ubuntu. It is wrong to say that this message can be ignored.
I am just copying from https://blaze-tutorial.meteor.com . I had progressed till almost half the tutorial. Then the problem started.
Now I have deleted and recreated the project many times over. No sooner than I change main.html, the error message starts appearing and localhost:3000 doesn't reflect ANY change whatsoever.

I am sorry to say, I am bitter with my experience with Meteor for unending compatibility issues. My Capstone project on Coursera is stuck because of such troubles for months. I wish I had invested my time in something less sophisticated but something that could function.

**Comment by tomtom87** [comment]
I have changed ddp client and server and there is no change at all. Also it reverts to the previous versions no matter what I specify in the versions file. I didn't update anything this was just pushed out afaik one minute it was fine, next this message appears.

**Comment by filipenevola** [comment]
Hi @tomtom87 so you can reproduce in latest versions?

Could you paste the content of your .meteor/packages, .meteor/versions and .meteor/release files?

**Comment by bhushitjoshipura** [comment]
I uninstalled and reinstalled Meteor. The problem seems to have gone for now.

**Comment by tomtom87** [comment]
release: ***@***.***

I am literally editing the versions files and setting the ddp client and
server @ versions and everytime it reverts to the originals. Is there
another way to set this or what am I doing wrong?

**Comment by filipenevola** [comment]
@tomtom87 you should not edit the `versions` file manually, this doesn't work, Meteor is always going to use the pinned version from the Meteor release for core packages.

For core packages, you need to be in the correct Meteor release to be able to run the correct package or you can fork and copy the package to your local folder `packages` which I believe makes no sense in this case.

For other packages, you could pin the version in the `packages` file but not for core packages.

And you should never change in the `versions` file.

**Comment by tomtom87** [comment]
How do I fix the issue? Update meteor? Please give a plain English business answer, thanks for all the verbose details, still doesn’t fix the problem. If it even is a problem. It’s just kinda weird to see it appear in my app, I never realized you can push our updates without our consent. Having to seriously think about Meteor now and going forward with it in production. 

>

**Comment by StorytellerCZ** [comment]
@tomtom87 You fix the issue by running `meteor update` and getting the latest version of the packages. The display of the message does not have any functional impact. Meteor does not and cannot push any updates without you expressly running a command to do so.

**Comment by tomtom87** [comment]
How come this error appears for everyone at the same time yesterday and non of us did an update?

>

**Comment by StorytellerCZ** [comment]
@tomtom87 It did not appear to everyone at the same time. If you look this issue has been started a while ago. We have fixed an issue where this message would be displayed every time on a start of a server as a check for this particular handshake between server and client was removed in a deprecated code.
That does not mean that the message might not appear when there is an actual invalid message that is not recognized by DDP. If you are on the latest version of `ddp-client` and `ddp-server` there might be something else going on and I first I would check what changes were done in your app just before the message started appearing in your app if something doesn't pop up, especially if you are working with DDP directly. That might at least give us an idea of what is going on with your app specifically.

**Comment by tomtom87** [comment]
It just appeared yesterday. Didn’t update anything. Started my dev env and this message is there.

I didn’t realize you can just roll out updates to meteor like that. I thought it was down to me to issue an update. Thanks for clarifying. 

>

**Comment by StorytellerCZ** [comment]
@tomtom87 We can't roll out any updates to your app. You always have to issue a command.

**Comment by tomtom87** [comment]
I started my app yesterday, there is this livedata message showing. I told I can’t get rid of it without updating. It wasn’t there Monday. It’s there Tuesday. I’ve done no update. So that doesn’t make any sense at all. 

>

**Comment by StorytellerCZ** [comment]
@tomtom87  As mentioned before, there can be multiple causes for the message. Without more information about your app I can only make wild guesses.

**Comment by tomtom87** [comment]
There’s seriously so many reasons for on one day having no message, to the next having the message without doing any update? Ok shoot man, take a wild guess, I’d love to hear it. 🍿 

>

**Comment by StorytellerCZ** [comment]
If you are on Meteor 2.3.5 with the latest packages then the problem you are seeing is not related to this issue and is something else. If you are confident that it is not coming from something in you app, please open a new issue with reproduction so that we can look into that. Wild guessing, without even knowing what version of Meteor and packages you are on, is just a waste of time.

**Comment by kakadais** [comment]
Hey guys- take it easy- ;)

I believe that what the @tomtom87 wants to know is 'how come it happened.
He's not saying 'it's real problem' or so on.

I had the same things before I saw @StorytellerCZ's comment,
because I love so much that the independence,
which mean,
'If I keep the project, it will work perfectly same in the 1000 years later'

I realize this is not true because,
Meteor is updating on its own even we could run the old project with the same codes & packages.

$(home)/.meteor has each version of Meteor build system and is treating all versions of Meteor projects,
but the core or some part could be handled by MDG update in some way that I don't understand for now.
I just believed too much the System would be perfect to be independent of the next changes. ;)

I don't think I understand fully for this problem,
but at least I understand that some part of the Meteor build system could be changed or controlled without the project changes,
and I think the reason I didn't get it is, 
the maintenance guys of Meteor are so good at avoiding this problem so far.

What a great job. ;) 
So what we need to do as a user of Meteor is just 'claps'.

But in spite of this understanding,
I want to say that if we can make this more 'Independently' that would be better,
even if this is enough and better than another platform for now.

people want to run their project exactly the same if they don't change anything, even 1000 years later ;)

Thanks all guys for your work again-

**Comment by tomtom87** [comment]
Don’t worry I’m chill man I just don’t get how it can update itself. Traditionally it should be user invoked not pushed. The 1000 year analogy is a good example. It makes me uneasy that updates are pushed without our consent or any notification. I just loaded my app that’s all. I did not issue a update command.

Meteor version 2.0

>

**Comment by kakadais** [comment]
emm, yeah it's curious ;)

@StorytellerCZ help man-
I think this is the last question haha.

![image](https://user-images.githubusercontent.com/5265343/130238365-17773ff6-2c1e-4799-871e-815b48e1298e.png)


I have the Linux server which is installed only 1.10.2 and never update anything on meteor & project here,
but the log message is showed when I re-run the project.

How the log message update is coming through in this project?
Did I have something to do on this project or just running makes something update?

**Comment by alihuber** [comment]
I'm also seeing this message appear on my dev machines all of a sudden. Meteor versions ranging from 1.11 (ddp-client 2.3.3) to 2.0 (ddp-client 2.4.0). Also without any manual intervention or updating anything. One of the apps was last updated in September 2020.
What makes this kinda disturbing is that it seems that somehow the server code on our (dev-)machines or even Meteor itself got updated without manually triggering an update. Something had to change to make this messages appear, or am I wrong? Especially when one first tries to exlain it with changed behaviour in Meteor 2.3, or ddp-client being out of date. Out of date compared to what? None of this is the case, it was running like this for months.
The other plausible reason is of course that there was a change in any Meteor package in the past that changed the behaviour of the log message, and for some reason this behaviour wasn't triggered in all of our apps until last month.

**Comment by StorytellerCZ** [comment]
> And BTW, this sentence from Storyteller was not correct:
>
> >@tomtom87 We can't roll out any updates to your app. You always have to issue a command.

Semantics, but true, mea culpa. What I meant there, there is no button the Meteor Software can push in some administration that is going to push code that is going to send code to user's apps.

**Comment by alihuber** [comment]
Thanks for clarification.

**Comment by zodern** [comment]
I also started recently seeing this error on an app still using Meteor 2.0.

It uses `ddp-client@2.4.0`, updated 7 months ago and `ddp-server@2.3.2`, updated 12 months ago. The versions for the core Meteor packages haven't changed in the past 4 months in this app. I have local packages disabled for this app (`METEOR_PACKAGE_DIRS` is set to an empty string).

Is there anything that could cause this error to start appearing besides the version of `ddp-client` or `dd-server` changing?

**Comment by kakadais** [comment]
Wow. 
The only question left is how I didn't know this for 7 years with 100 projects ;)
Thanks for clarification too-

**Comment by filipenevola** [comment]
@kakadais we don't have that many patch versions in core packages.

Usually we have only minor versions bumps on each Meteor release. Also it's not very often that something is going to produce a new log in every new server startup and so in this case it was very clear that something was changed.

**Comment by tomtom87** [comment]
Exactly 

>

**Comment by StorytellerCZ** [comment]
@zodern 
This is the offending code in ddp-client:
```javascript
if (msg === null || !msg.msg) {
      if (!msg || !msg.testMessageOnConnect) {
        if (Object.keys(msg).length === 1 && msg.server_id) return; // added this to identify and ignore the message that originally triggered the message

        Meteor._debug('discarding invalid livedata message', msg);
      }

      return;
}
```
So it can be triggered in some other instances. Also don't mistake this for `discarding unknown livedata message type` which is when ddp can't figure out what to do with the received message.

**Comment by zodern** [comment]
@StorytellerCZ  The error log is:
```
discarding invalid livedata message { testMessageOnConnect: true }
```

I actually am seeing the error message before the app's server starts. Could the invalid message be sent by troposphere when Meteor is updating the package catalog?

**Comment by filipenevola** [comment]
I believe we are spending too much time here but I can explain again.

This message was there from Meteor versions before 1.0.

Before 1.0 Meteor was sending a message after the Websocket connection.

This message was used in some tests to check if Fibers was running in the callback and by mistake this message was removed by Storyteller when he was removing old code.

So I added this message back but with this boolean `testMessageOnConnect` so we could ignore this message in the client and avoiding this log.

By a mismatch in the versions of ddp-client and ddp-server maybe the message will still show up as one package is sending one message but the other package is not checking for this message to avoid the log.

But again, this code is not affecting anything, it was an old pattern that is useful only in tests.
